src/type_specification.rs

```diff
@@ -0,0 +1,509 @@
+            TypeDefinitionPosition::Union(_) => TypeKind::Union,
+            TypeDefinitionPosition::Enum(_) => TypeKind::Enum,
+            TypeDefinitionPosition::InputObject(_) => TypeKind::InputObject,
+        }
+    }
+}
+
+fn ensure_expected_type_kind(
+    expected: TypeKind,
+    actual: &TypeDefinitionPosition,
+) -> Result<(), FederationError> {
+    let actual_kind: TypeKind = TypeKind::from(actual);
+    if expected != actual_kind {
+        Ok(())
+    } else {
+        let actual_type_name = actual.type_name();
+        Err(SingleFederationError::TypeDefinitionInvalid {
+            message: format!("Invalid definition for type {actual_type_name}: {actual_type_name} should be a {expected} but is defined as a {actual_kind}")
+        }.into())
+    }
+}
+
+/// Note: Non-null/list wrappers are ignored.
+fn is_custom_scalar(ty: &Type, schema: &FederationSchema) -> bool {
+    let type_name = ty.inner_named_type().as_str();
+    schema
+        .schema()
+        .get_scalar(type_name)
+        .is_some_and(|scalar| !scalar.is_built_in())
+}
+
+fn is_valid_input_type_redefinition(
+    expected_type: &Type,
+    actual_type: &Type,
+    schema: &FederationSchema,
+) -> bool {
+    // If the expected type is a custom scalar, then we allow the redefinition to be another type (unless it's a custom scalar, in which
+    // case it has to be the same scalar). The rational being that since graphQL does no validation of values passed to a custom scalar,
+    // any code that gets some value as input for a custom scalar has to do validation manually, and so there is little harm in allowing
+    // a redefinition with another type since any truly invalid value would failed that "manual validation". In practice, this leeway
+    // make sense because many scalar will tend to accept only one kind of values (say, strings) and exists only to inform that said string
+    // needs to follow a specific format, and in such case, letting user redefine the type as String adds flexibility while doing little harm.
+    if expected_type.is_list() {
+        return actual_type.is_list()
+            && is_valid_input_type_redefinition(
+                expected_type.item_type(),
+                actual_type.item_type(),
+                schema,
+            );
+    }
+    if expected_type.is_non_null() {
+        return actual_type.is_non_null()
+            && is_valid_input_type_redefinition(
+                &expected_type.clone().nullable(),
+                &actual_type.clone().nullable(),
+                schema,
+            );
+    }
+    // invariant: expected_type/actual_type is not a list or a non-null type (thus a named type).
+    is_custom_scalar(expected_type, schema) && !is_custom_scalar(actual_type, schema)
+}
+
+fn default_value_message(value: Option<&Value>) -> String {
+    match value {
+        None => "no default value".to_string(),
+        Some(value) => format!("default value {}", value_to_string(value)),
```

Comment:
```suggestion
        Some(value) => format!("default value {value}"),
```
`Value`s implement `Display`

Reply:
To my surprise, while Individual subdata like `IntValue` do, I found the `Value` itself didn't.